### PR TITLE
add a plugin to trace react updates 

### PR DESCRIPTION
**Pull request by hedgerwang**
Add a plugin (codename: bananaslug) to highlight react updates in page. 
See https://github.com/hedgerwang/react-devtools/tree/banana_slug/demo for demo and details.

### CODE REVIEW

**Comment by jimfb** [comment]
@spicyj This is a very cool feature, we should figure out how to get it merged.  Since you know the devtools best, can you take a look?

**Comment by hedgerwang** [comment]
cc @sebmarkbage @jordwalke

**Comment by spicyj** [comment]
Do you think you could take a look at this @jaredly?

**Comment by jaredly** [comment]
sure thing

File: plugins/BananaSlug/BananaSlugWebNodeMeasurer.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  height: 0,
+  left: 0,
+  right: 0,
+  scrollX: 0,
+  scrollY: 0,
+  top: 0,
+  width: 0,
+};
+
+class BananaSlugWebNodeMeasurer extends BananaSlugAbstractNodeMeasurer {
```
Comment (jaredly) [inline plugins/BananaSlug/BananaSlugWebNodeMeasurer.js]: is this class hierarchy useful?

**Comment by jaredly** [comment]
How does this do perf-wise?

File: plugins/BananaSlug/BananaSlugWebNodeMeasurer.js (hunk 1 above)
Comment (hedgerwang) [inline plugins/BananaSlug/BananaSlugWebNodeMeasurer.js]: It makes sense to me.

I want to encapsulate the measurement logic within platform-specific modules.

For example, `BananaSlugWebNodeMeasurer` implements the method to measure a web node. I'd imagine that we can add `BananaSlugIOSViewMeasurer` to measure a view as well.

**Comment by hedgerwang** [comment]
@jaredly : Both the measurement and painting are throttled and cached. The feature can be turned off so I don't see anything major to slow down the target app. 

I tested it and it seems working fine for me.

File: plugins/BananaSlug/BananaSlugWebNodeMeasurer.js (hunk 1 above)
Comment (jaredly) [inline plugins/BananaSlug/BananaSlugWebNodeMeasurer.js]: certainly separating the logic is great, but composition would work just as well, right? class hierarchies tend to be harder to maintain (imo), and especially so in a non-statically typed language.

speaking of which -- could you get @flow annotations on this PR?

**Comment by jaredly** [comment]
Cool. yeah, the throttling + caching stuff looks great

**Comment by hedgerwang** [comment]
@jaredly : will `npm lint` also performs `flow` checks? If not, what's the command that I shall use to test this?

Thanks.

**Comment by kassens** [comment]
@hedgerwang If you have `flow` installed on your computer you should be able to just run `flow` in the main directory. Otherwise `node_modules/.bin/flow` should do the trick. (flow also runs on Travis, so it should be all good)

File: plugins/BananaSlug/BananaSlugWebNodeMeasurer.js (hunk 1 above)
Comment (hedgerwang) [inline plugins/BananaSlug/BananaSlugWebNodeMeasurer.js]: I had add **@flow** to enforce the interface, so OO-style interface is enforced :)

**Comment by hedgerwang** [comment]
@jaredly : flow check was added. please have another look. thanks.

**Comment by hedgerwang** [comment]
ping :)

File: plugins/BananaSlug/BananaSlugFrontendControl.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+type DefaultProps = {};
+
+const StateRecord = immutable.Record({
+  enabled: false,
+});
+
+class BananaSlugFrontendControl extends
+  React.Component<DefaultProps, Props, State> {
+  _defaultState: ControlState;
+  _toogle: (b: boolean) => void;
```
Comment (jaredly) [inline plugins/BananaSlug/BananaSlugFrontendControl.js]: `_toggle`?

**Comment by jaredly** [comment]
This is so awsome!
could you remove the `demo` directory, and add firefox support? then I think it'll be ready to go

**Comment by hedgerwang** [comment]
@jaredly 

I've removed the `demo` directory and added the firefox support.
Please have another look :)

**Comment by jaredly** [comment]
looks like it needs a rebase, and then it's great!
@jimfb after the rebase, we should be all set

**Comment by jimfb** [comment]
Ok, great, thanks @jaredly!

**Comment by hedgerwang** [comment]
@jimfb, @jaredly : I've done the rebase work. Let's ship it!

**Comment by hedgerwang** [comment]
happy new year!
any chance that we could land this?

**Comment by sebmarkbage** [comment]
Did you test it with React Native so the UI doesn't break there? In other news, does the devtools work in React Native again? Anyway. I'll merge and then we'll see what happens.

**Comment by spicyj** [comment]
> In other news, does the devtools work in React Native again?

No. :( #229

**Comment by hedgerwang** [comment]
@sebmarkbage :  there is no implementation in RN so this feature is disabled in RN.

**Comment by hedgerwang** [comment]
Thanks everyone who helped to review this. 

One more question:
Is there anything that I should do to land the new extension to web store?

Thanks.

**Comment by huang47** [comment]
this is awesome, just can't wait! :D

**Comment by gaearon** [comment]
Every time I read the source I forget what "banana slug" means.
😞

**Comment by spicyj** [comment]
@gaearon Feel free to rename. :) RerenderIndicator

**Comment by hedgerwang** [comment]
@gaearon : Banaslug is a lovely creature that moves slowly and leaves visible trace on the ground.

image

**Pull request merged**